### ISSUE TITLE
SqlConnection should support Azure AD token-based authentication

### ISSUE THREAD

**Issue by nathandavidson** [dotnet/corefx#13660]
The desktop .NET Framework 4.6 and newer has an `AccessToken` property on the `SqlConnection` class ([MSDN](https://msdn.microsoft.com/en-us/library/system.data.sqlclient.sqlconnection.accesstoken(v=vs.110).aspx)) which can be used to authenticate to a SQL Azure database using an access token issued by Azure AD ([examples here](https://azure.microsoft.com/en-us/documentation/articles/sql-database-aad-authentication/#7-connect-by-using-azure-ad-identities)). However, this property is not present on the version of `SqlConnection` provided in the System.Data.SqlClient NuGet package, including the latest preview v4.3.0-preview1-24530-04. In case it's an important detail, the project in question is a library project targeting 'netstandard1.6'.

The lack of the `AccessToken` property makes it difficult, if not impossible, to support modern directory-based auth scenarios for Azure SQL DB while also targeting .NET Standard (I have not attempted to use the connection string-based AAD options, but for various reasons those would not be applicable for my project). I have yet to find a workaround other than setting the target framework to 'net46', which I would like to avoid except as a last resort.

**Comment by saurabh500** [comment]
Related to #8807

**Comment by stijnherreman** [comment]
It's a bit strange that Azure is such a huge and important part of Microsoft, but .NET Standard and .NET Core aren't given the necessary resources to support it.

I don't mean to blame anyone here, it's just that I'd love to use .NET Core but still cannot do so more than 3 years after it was announced.

**Comment by stijnherreman** [comment]
The PR referred to by @keeratsingh is https://github.com/dotnet/corefx/pull/30342

**Comment by AfsanehR** [comment]
@nathandavidson @stijnherreman Closing this is issue since it is addressed by pr [#30342](https://github.com/dotnet/corefx/pull/30342).

**Comment by Liero** [comment]
@AfsanehR: great! So this will be released in .NET Core 2.2?

**Comment by BruceHunter** [comment]
Security should be number 1. This should have been released already.

**Comment by karelz** [comment]
The PR was merged into master, which will ship in 3.0.
I let @AfsanehR @David-Engel @keeratsingh to comment on option porting it to 2.2.

**Comment by stijnherreman** [comment]
I think everyone waiting for this would like to see it in 2.2.

**Comment by Trontin** [comment]
How this implementation will impact the pool connection key ? If there is two differents AccessToken with the same connection string, does it mean that they can share the same connection pool or not ?

**Comment by saurabh500** [comment]
A new connection pool is created for every access token.

**Comment by paulirwin** [comment]
+1 for getting this into .NET Core 2.2. By the time 2.2 comes out this issue will be 2 years old.

**Comment by wienleung** [comment]
@AfsanehR  @David-Engel @keeratsingh any decision on whether this update will be ported to 2.2?

**Comment by keeratsingh** [comment]
@wienleung @Liero @paulirwin @stijnherreman We are currently in the process of porting this fix over to .NET Core 2.2, will updated here once the process is complete.

**Comment by karelz** [comment]
Reopening the issue to track port to 2.2.

**Comment by AfsanehR** [comment]
The AccessToken property is now added to .NET Core 2.2.  🎉

**Comment by danmosemsft** [comment]
Thanks @AfsanehR for addressing this popular issue. For those who will ask: we have not AFAIK announced a date for 2.2 yet. It will definitely be before 3.0 😀 Meantime, if anyone has a chance to try this out of the repo (either master or release/2.2) it would be great if you could indicate here to confirm you were successful.

**Comment by karelz** [comment]
2.2. is currently planned for Q4 2018: https://github.com/dotnet/core/blob/master/roadmap.md#upcoming-ship-dates

**Comment by bloodybeet** [comment]
Can you give us an update on when we should be able to test this feature/start using it in development. Would that be 2.2.100-preview2?

**Comment by jnevins-gcm** [comment]
Why can't this be distributed as a nuget package as opposed to with the full core sdk?

**Comment by karelz** [comment]
@jnevins-gcm shipping standalone packages leads to test matrix explosion + things don't work properly together when they should, etc. It caused lots of troubles for both the customers and the team and with .NET Core we started shipping platform as single NuGet package.
You can check the feature in daily builds of release/2.2 build, or wait for next preview which will contain the fix (based on times and the schedule I am aware of, it should be Preview2).

**Comment by jnevins-gcm** [comment]
Can we consume 2.2 as a nuget package instead of having to use the windows installer for the core sdk?

**Comment by karelz** [comment]
@jnevins-gcm I think you will need the SDK, but I let area owners to have final word on the topic.

**Comment by jnevins-gcm** [comment]
Ok, is this the right forum to get that answer? Who should I be asking?

**Comment by AfsanehR** [comment]
@jnevins-gcm @karelz is correct. You need to update the SDK and SqlClient version both. 
I was able to test it with SDK version 2.2.0-preview-26820-02 and SqlClient version 4.6.0-preview2-26905-02

**Comment by jnevins-gcm** [comment]
The sqlclient is included in the sdk’s net ore reference assemblies though?

**Comment by jnevins-gcm** [comment]
If the sqlclient update is separate, where do I get it?

**Comment by karelz** [comment]
@jnevins-gcm SlqClient update is NOT separate. We do not ship or support separate packages which are part of .NET Core platform -- see my previous reply https://github.com/dotnet/corefx/issues/13660#issuecomment-420337898 explaining why.

**Comment by AfsanehR** [comment]
It is included but you need to include the version you want in your csproj file. 
For example:
```
<ItemGroup>
    <PackageReference Include="System.Data.SqlClient" Version="4.6.0-preview2-26905-02" />
  </ItemGroup>
```
You can check the latest version of SqlClient from here:
https://dotnet.myget.org/feed/dotnet-core/package/nuget/System.Data.SqlClient

**Comment by jnevins-gcm** [comment]
I guess I don’t understand your release model. 

netstandard2.0 started down the road of “no individual nuget packages for libraries that are part of the netstandard surface area”. You reference netstandard (implicitly) and the platform specific netstandard implementation forwards to the correct platform specific (or platform agnostic) implementation assembly that is distributed as part of that platform’s clr redistributable. 

SqlClient isn’t part of netstandard. 

Okay, so that being the case, how is it tied to a specific version of the coreclr? I’m looking at your nuget package on myget. It has both platform neutral implementations and a netcoreapp specific implementation. But no netcoreapp2.2 specific implementation. 

Can you please help me understand how SqlClient is coupled to the netcore sdk and where that dependency is expressed and why I can’t or shouldn’t be able to just add a PackageReference to the one you linked to on myget without installing netcore sdk 2.2?

**Comment by swettstein** [comment]
according to the readme, this should be included in [preview1](https://github.com/dotnet/core/blob/v2.2.0-preview1/release-notes/2.2/preview/2.2.0-preview1.md) but after i installed the SDK and added the new Sql package to version 4.6.0-preview2-26905-02 but still don't see the AccessToken on SqlConnection

**Comment by karelz** [comment]
@swettstein while the PR was merged for 2.2 Preview1, it omitted packaging changes leading to not shipping the bits at all. They will be available in 2.2 Preview2 (also clarified above - https://github.com/dotnet/corefx/issues/13660#issuecomment-413921458).
I checked with experts and this particular package is shipping also standalone, so it should be applicable to older version (although we do not test the combination).

**Comment by jnevins-gcm** [comment]
So is that to say my understanding as I described above is actually correct?

**Comment by karelz** [comment]
@jnevins-gcm (had to check with experts again)
All APIs which are part of netstandard has to be implemented in each platform (.NET Core, Mono, .NET Framework, UWP) that supports that version of .NET Standard inbox (without external packages).
There are APIs which are in the platform, but do not ship as individual packages.
SqlClient already shipped as individual package before and we kept it shipping that way, that's why it ships standalone.

In nutshell, whatever is part of netstandard, has to be in each platform inbox. Some APIs (outside of netstandard) may ship as well as standalone packages. Decision what ships as standalone package is based on business need, and not on "can it be done".

Hope it clarifies things. Let's keep the general platform/targeting discussion separate. It is not specific to SqlClient.

**Comment by bloodybeet** [comment]
My use case is as follows:
- I created an Azure Service Principal and added it to AD Group X
- Made AD Group X the "Active Directory Admin" of SQL Server Y

`var context = new AuthenticationContext($"https://login.microsoftonline.com/{tenantId}");
            var result = await context.AcquireTokenAsync("https://management.core.windows.net/", new ClientCredential(configuration["ServicePrincipalAppId"], configuration["ServicePrincipalAppSecret"]));
using (var conn = new SqlConnection("Data Source=servery.database.windows.net"))
            {
                conn.AccessToken = result.AccessToken;
                conn.Open();
            }`

This gives me a "'Login failed for user 'NT AUTHORITY\ANONYMOUS LOGON'.'"

Is this supposed to work with Service Principals as well?

**Comment by karelz** [comment]
@bloodybeet did you use [2.2 Preview2](https://github.com/dotnet/core/blob/master/release-notes/2.2/preview/2.2.0-preview2.md)?
If you found a limitation in functionality, I would recommend to file a new issue to avoid mixing up multiple problems and discussions here.

**Comment by bloodybeet** [comment]
Yes I am using 4.6.0-preview2-26905-02 - is this the latest one?

**Comment by AfsanehR** [comment]
@bloodybeet Would you be able to file a new issue as Karel recommended? It would be easier to continue discussion there. In the meantime, I would recommend to check if you have created a contained database user and grant the necessary permissions. Please refer to this link for more information
https://docs.microsoft.com/en-us/azure/sql-database/sql-database-aad-authentication-configure

**Comment by bloodybeet** [comment]
Even better: the access token wasn't coming from database.windows.net, but management.core.windows.net - fixed and working as expected. I have deleted all messages to keep this thread clean ;)
